Move worker name collection code from Python to C++

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25499 [WIP] Add Python RRef type
* #25169 Adding RRef as return value for builtin operators
* **#24260 Move worker name collection code from Python to C++**
* #24195 Assign each RpcAgent a unique ID, and use ID for sending RPC messages.

This also simplifies ProcessGroupAgent constructor signature.

Differential Revision: [D16789219](https://our.internmc.facebook.com/intern/diff/D16789219)

mrshenli committed:
Move worker name collection code from Python to C++

This also simplifies ProcessGroupAgent constructor signature.

pritamdamania87 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -53,17 +53,66 @@ Message deserialize(std::istream& is) {
 }
 
+std::unordered_map<std::string, int> collectNames(
+    const std::string& workerName, std::shared_ptr<c10d::ProcessGroup>& pg) {
```
Should we have some restrictions on the workerName? For ex: no non-ASCII chars, no special chars etc. A simple rule would be to just allow [A-Za-z0-9] and not allowing the first char to be a number.

mrshenli replied:
Sure, I will add restrictions on worker names.

pritamdamania87 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -53,17 +53,66 @@ Message deserialize(std::istream& is) {
       stop_(false),
       pg_(std::move(pg)),
       nextId_(0) {
+  nameMap_ = collectNames(workerName_, pg_);
```
nit: can move this into the constructor initializer list above.

mrshenli replied:
good point, let me do that.

pritamdamania87 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -53,17 +53,66 @@ Message deserialize(std::istream& is) {
   return Message(std::move(payload), std::move(tensors), type, id);
 }
 
+std::unordered_map<std::string, int> collectNames(
```
Why don't we make this a private member of ProcessGroupAgent since we need pg?

mrshenli replied:
Will make the change.

pritamdamania87 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -53,17 +53,66 @@ Message deserialize(std::istream& is) {
+std::unordered_map<std::string, int> collectNames(
+    const std::string& workerName, std::shared_ptr<c10d::ProcessGroup>& pg) {
+
+  const auto worldSize = pg->getSize();
```
I think we can simplify this function if we enforce a limit on the max length of workerName (ex: 128/256 chars). With this, we can just initialize tensors with this max length and just do one all gather to collect all the worker names. All the worker names would be null terminated and hence we don't even need to collect their lengths first. std::string() would do the right thing and just read up to the first null terminated character when we just pass in char * with no length.

mrshenli replied:
Sure, but all the optimization we do here will probably gone after we address #23232, as we will call into that instead. Do you still prefer adding the temporary optimization here? LMK, I will make the change.

pritamdamania87 replied:
I think let's just add the temporary optimization for now since its not clear when we might address https://github.com/pytorch/pytorch/issues/23232. Also, apart from an optimization, it also just makes the code a lot simpler and cleaner.

pritamdamania87 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -53,17 +53,66 @@ Message deserialize(std::istream& is) {
+    TORCH_CHECK(nameMap.find(peerName) == nameMap.end(),
+        "RpcAgent name ", peerName, " is not unique.");
+
+    nameMap[std::move(peerName)] = i;
```
Does all_gather guarantee that `outputNames` will be populated in the order of ranks?

mrshenli replied:
Yes, it should.

mrshenli committed:
Update on "Move worker name collection code from Python to C++"


This also simplifies ProcessGroupAgent constructor signature.

Differential Revision: [D16789219](https://our.internmc.facebook.com/intern/diff/D16789219)

mrshenli committed:
Update on "Move worker name collection code from Python to C++"


This also simplifies ProcessGroupAgent constructor signature.

Differential Revision: [D16789219](https://our.internmc.facebook.com/intern/diff/D16789219)

mrshenli committed:
Update on "Move worker name collection code from Python to C++"


This also simplifies ProcessGroupAgent constructor signature.

Differential Revision: [D16789219](https://our.internmc.facebook.com/intern/diff/D16789219)

mrshenli committed:
Update on "Move worker name collection code from Python to C++"


This also simplifies ProcessGroupAgent constructor signature.

Differential Revision: [D16789219](https://our.internmc.facebook.com/intern/diff/D16789219)

mrshenli committed:
Update on "Move worker name collection code from Python to C++"


This also simplifies ProcessGroupAgent constructor signature.

Differential Revision: [D16789219](https://our.internmc.facebook.com/intern/diff/D16789219)

mrshenli committed:
Update on "Move worker name collection code from Python to C++"


This also simplifies ProcessGroupAgent constructor signature.

Differential Revision: [D16789219](https://our.internmc.facebook.com/intern/diff/D16789219)

mrshenli committed:
Update on "Move worker name collection code from Python to C++"


This also simplifies ProcessGroupAgent constructor signature.

Differential Revision: [D16789219](https://our.internmc.facebook.com/intern/diff/D16789219)

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -55,21 +55,52 @@ Message deserialize(MessageType type, std::istream& is) {
+  // use c10d allgather to collect names
+  torch::Tensor nameTensor =
+      torch::zeros({WorkerId::MAX_NAME_LEN}, torch::kChar);
+  memcpy(nameTensor.storage().data(), workerName.c_str(), length);
```
nit: can we just use `workerName.length()` here and get rid of the `length` variable?

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -55,21 +55,52 @@ Message deserialize(MessageType type, std::istream& is) {
+  memcpy(nameTensor.storage().data(), workerName.c_str(), length);
+  std::vector<torch::Tensor> inputName = {nameTensor};
+  std::vector<std::vector<torch::Tensor>> outputNames(1);
+  for (int i = 0; i < worldSize; ++i) {
```
nit: size_t

mrshenli replied:
why size_t here? `pg_->getSize()` returns an int.

pritamdamania87 replied:
Oops didn't realize it returns `size`, usually the convention in C++ is to use `size_t` for sizes.

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -55,21 +55,52 @@ Message deserialize(MessageType type, std::istream& is) {
+  pg_->allgather(outputNames, inputName)->wait();
+
+  // convert collected name tensors into string names
+  for (int i = 0; i < worldSize; ++i) {
```
nit: size_t

mrshenli replied:
ditto

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -4,6 +4,8 @@ namespace torch {
 namespace distributed {
 namespace rpc {
 
+const size_t WorkerId::MAX_NAME_LEN=128;
```
nit: can use contexpr inline in the header file.

mrshenli replied:
cannot use inline var though, that's C++17

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -12,13 +14,36 @@ using worker_id_t = int16_t;
+      : name_(std::move(name)), id_(id) {
+    bool validSize = name_.length() < MAX_NAME_LEN && name_.length() > 0;
+    bool validChar =
+        find_if(
```
Shouldn't this be std::find_if?

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -12,13 +14,36 @@ using worker_id_t = int16_t;
+            name_.begin(),
+            name_.end(),
+            [](char c) {
+              return !(isalnum(c) || c == '-' || c == '_') ;
```
std::isalnum?

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -12,13 +14,36 @@ using worker_id_t = int16_t;
+              return !(isalnum(c) || c == '-' || c == '_') ;
+            }
+        ) == name_.end();
+    TORCH_CHECK(validSize && validChar,
```
Should we allow worker names to start with a digit, "-" or "_"? Seems like worker name should always start with an alphabet (usually this is what is done in other systems).

mrshenli replied:
I'd prefer not over engineering this here. What benefits do we or users get if we put more restrictions here? It looks fine to me if users want to do things like `init_model_parallel("1")` or `init_model_parallel("_hidden_monitor")`?

pritamdamania87 replied:
I agree, let's keep this more flexible for now.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -12,13 +14,36 @@ using worker_id_t = int16_t;
+        "but got ", name_);
+  }
+
+  static bool isValidName(const std::string& name) {
```
This isn't used anywhere?

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -12,13 +14,36 @@ using worker_id_t = int16_t;
   WorkerId(std::string name, worker_id_t id)
-      : name_(std::move(name)), id_(id) {}
+      : name_(std::move(name)), id_(id) {
+    bool validSize = name_.length() < MAX_NAME_LEN && name_.length() > 0;
```
Can we also add unit tests which test various different combinations of valid and invalid worker names?

mrshenli replied:
Yes, will add.

mrshenli committed:
Update on "Move worker name collection code from Python to C++"


This also simplifies ProcessGroupAgent constructor signature.

Differential Revision: [D16789219](https://our.internmc.facebook.com/intern/diff/D16789219)

mrshenli committed:
Update on "Move worker name collection code from Python to C++"


This also simplifies ProcessGroupAgent constructor signature.

Differential Revision: [D16789219](https://our.internmc.facebook.com/intern/diff/D16789219)

mrshenli committed:
Update on "Move worker name collection code from Python to C++"


This also simplifies ProcessGroupAgent constructor signature.

Differential Revision: [D16789219](https://our.internmc.facebook.com/intern/diff/D16789219)

mrshenli committed:
Update on "Move worker name collection code from Python to C++"


This also simplifies ProcessGroupAgent constructor signature.

Differential Revision: [D16789219](https://our.internmc.facebook.com/intern/diff/D16789219)

mrshenli committed:
Update on "Move worker name collection code from Python to C++"


This also simplifies ProcessGroupAgent constructor signature.

Differential Revision: [D16789219](https://our.internmc.facebook.com/intern/diff/D16789219)

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -4,7 +4,7 @@ namespace torch {
 namespace rpc {
 
-const size_t WorkerId::MAX_NAME_LEN=128;
+constexpr size_t WorkerId::MAX_NAME_LEN;
```
I meant using only `constexpr` keyword but inside the header file. Didn't meant `constexpr inline`

pietern replied:
@mrshenli This is now defined both in the header and here. This one can go.
